File: 灵茶/动态规划/0377.java
import java.util.*;

class Solution {
    public int combinationSum4(int[] nums, int target) {
        int[] d = new int[target + 1];
        Arrays.sort(nums);
        d[0] = 1;
        for (int i = nums[0]; i <= target; i++) {
            for (int num : nums) {
                if (i >= num) {
                    d[i] += d[i - num];
                }
            }
        }
        return d[target];
    }
}

File: 灵茶/动态规划/0377.py
from functools import cache
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        @cache
        def dfs( i : int):
            if i == 0 :
                return 1
            return sum(dfs(i - x) for x in nums if x <= i) 
        return dfs(target)